jax/_src/nn/functions.py

```diff
@@ -765,3 +768,203 @@ def hard_silu(x: ArrayLike) -> Array:
   return x_arr * hard_sigmoid(x_arr)
 
 hard_swish = hard_silu
+
+def _causal_mask(T, dtype):
+  dtype_max = jnp.finfo(dtype).max
+  large_negative_number = jnp.asarray(-0.5 * dtype_max, dtype=dtype)
+  col_idx = jnp.tile(jnp.arange(T)[jnp.newaxis, :], [T, 1])
+  row_idx = jnp.tile(jnp.arange(T)[:, jnp.newaxis], [1, T])
+  mask = (row_idx < col_idx).astype(dtype) * large_negative_number
+  mask = mask[jnp.newaxis, jnp.newaxis, :, :]
+  return mask
+
+class SdpaCausalMask:
+  def __call__(self, x: ArrayLike, mask: ArrayLike | None = None) -> Array:
+    """Mask the input with causal masking"""
+    if mask is not None:
+      warnings.warn(
+          "The provided mask will be ignored since the causal mask is used")
+      del mask
+    T = x.shape[-2]
+    assert T == x.shape[-1], f'T should equal to S: {T} vs {x.shape[-1]}'
+    assert jnp.issubdtype(x.dtype, jnp.floating), x.dtype
+    mask = _causal_mask(T, x.dtype)
+    return x + mask
+
+@dataclass
+class SdpaPhiloxDropout:
+  rate: float = 0.0
+  seed: int = 123
+  def __call__(self, x: ArrayLike) -> Array:
+    if self.rate == 0.0:
+      return x
+    else:
+      raise ValueError(
+          "This dropout should only be used in the context of cuDNN flash"
+          "attention; Please check if other spda configs are compatible.")
+
+def scaled_dot_product_attention(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None = None,
+    mask: ArrayLike | None = None,
+    *,
+    scale: float | None = None,
+    logits_cap: float | None = None,
+    mask_fn: Callable[[ArrayLike, ArrayLike], ArrayLike] = None,
```

Comment:
Do we have an application for the mask function beyond causal masking? This is such a common use-case that I think it should have its own option (`is_causal`. PyTorch has this as well https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html). Its ugly to have to pass `mask_fn=jax.nn.SdpaCausalMask()`.

Reply:
Masking is one way to implement causal attention, and good implementations avoid implementing it with masking! So it is conceptually distinct from masking, which should be reflected in the API.

Reply:
For the other masking types: I notice they are not supported in your current API yet. I think we should discuss these in a separate PR. But we should get rid of masking_fn for now.